[HUDI-3103] Enable MultiTableDeltaStreamer to update a single sink table from multiple source tables

watermelon12138 opened the pull request:
## What is the purpose of the pull request

*Enable MultiTableDeltaStreamer to update a single target table from multiple source tables.*

## Brief change log
  - *Modify the HoodieMultiTableDeltaStreamer file so that it can generate the execution context of table based on source tables.*
  - *Modify the DeltaSync.java file so that the source table can associate with other tables and the source can configure independent checkpoint.*
  - *add UT.*

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -388,6 +388,9 @@ private boolean onDeltaSyncShutdown(boolean error) {
     public Boolean retryLastPendingInlineClusteringJob = false;
 
+    @Parameter(names = {"--fetch-from-multi-sources"}, description = "Fetch data from multi sources and ingest to single sink table")
+    public Boolean isFetchFromMultiSources = false;
```
rename to allowFetchFromMultipleSources?

watermelon12138 replied:
ok, I've modified it.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -388,6 +388,9 @@ private boolean onDeltaSyncShutdown(boolean error) {
     @Parameter(names = {"--retry-last-pending-inline-clustering", "-rc"}, description = "Retry last pending inline clustering plan before writing to sink.")
     public Boolean retryLastPendingInlineClusteringJob = false;
 
+    @Parameter(names = {"--fetch-from-multi-sources"}, description = "Fetch data from multi sources and ingest to single sink table")
```
nit: in the description, lets change multi to multiple.

watermelon12138 replied:
ok, I've modified it.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieMultiTableDeltaStreamer.java:
```diff
@@ -61,8 +62,11 @@
   private static Logger logger = LogManager.getLogger(HoodieMultiTableDeltaStreamer.class);
 
   private List<TableExecutionContext> tableExecutionContexts;
+
```
let us remove these extra lines. :)

watermelon12138 replied:
ok, I've modified it.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieMultiTableDeltaStreamer.java:
```diff
@@ -355,6 +414,12 @@ public static void main(String[] args) throws IOException {
+     * Fetch data from multi sources.
+     */
+    @Parameter(names = {"--fetch-from-multi-sources"}, description = "Fetch data from multi sources and ingest to single sink table")
+    public Boolean isFetchFromMultiSources = false;
```
ditto, update the variable name and description

watermelon12138 replied:
ok, I've modified it.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieMultiTableDeltaStreamer.java:
```diff
@@ -379,23 +444,50 @@ private static String resetTarget(Config configuration, String database, String
+      return;
+    }
+    if (tableExecutionContexts.get(0).getConfig().isFetchFromMultiSources) {
+      while (true) {
```
This essentially makes it an ever running job. I guess we do not want users to have resources up and running always doing ingestion with no option of having a one time job. From what I understand you are trying to introduce a continuous mode here, which is not recommended this way. Let us make it configurable? cc @nsivabalan

pratyakshsharma replied:
Also trying to understand this scenario. Suppose a user has a combination of target tables to be ingested, say table A and table B. Table A is bound to multiple sources, but table B depends on a single source. I hope such combinations of tables are taken care. May be we can add test cases for such combinations if not done already? @watermelon12138

watermelon12138 replied:
Thank you for your good advice. First, I added the parameter allowContinuousWhenMultipleSources to control whether to use the continuous mode in the case of multiple sources. Second, In the original logic, I only allowed the sink table to be unique. However, I accept your suggestion. Currently, I support multiple sink tables, and each sink table can be bound to multiple sources.

pratyakshsharma reviewed (changes requested):
Thank you for raising this new PR @watermelon12138 . Few comments and clarifications. Once addressed, we should be good to land this. 

cc @bvaradar @nsivabalan , good to have your reviews as well on this

pratyakshsharma commented:
One high level question, is there any limitation or pre requisite in terms of different schemas from different source topics that users should be aware of when using this new feature? What happens when one source schema has few fields which are not present in second source schema and so on. Does that lead to data loss? How are we handling that?

watermelon12138 commented:
> One high level question, is there any limitation or pre requisite in terms of different schemas from different source topics that users should be aware of when using this new feature? What happens when one source schema has few fields which are not present in second source schema and so on. Does that lead to data loss? How are we handling that?
@pratyakshsharma 
Good question! This new feature is designed to support the service scenario in which multiple sources are injected to one sink table. When the schemas of multiple sources are inconsistent, we must configure the independent schema and transformer for each source to convert the schema of source to the schema of sink table so that source data can be written to the sink table. For example, We can configure hoodie.deltastreamer.schemaprovider.source.schema.file or hoodie.deltastreamer.source.schemaProviderClassName to specify the schema of each source, and then configure hoodie.deltastreamer.source.transformerClassNames or hoodie.deltastreamer.transformer.sql to convert the schema of source to the schema of sink table. I highly recommend the hoodie.deltastreamer.transformer.sql configuration, which can associate source data with Hive tables, for example, join. The preceding method is helpful for resolving schema inconsistency issues. I look forward to hearing from you.

pratyakshsharma commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieMultiTableDeltaStreamer.java:
```diff
@@ -177,6 +182,183 @@ public void testMultiTableExecutionWithKafkaSource() throws IOException {
+    String targetBasePath1 = executionContexts.get(0).getConfig().targetBasePath;
+    String targetBasePath2 = executionContexts.get(1).getConfig().targetBasePath;
+    streamer.sync();
+    // assert whether targetBasePath1 equals targetBasePath2 because that there is only one sink table.
```
remove that from comment

pratyakshsharma commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieMultiTableDeltaStreamer.java:
```diff
@@ -177,6 +182,183 @@ public void testMultiTableExecutionWithKafkaSource() throws IOException {
+  }
+
+  @Test
+  public void testMultiTableExecutionWithKafkaSourceWhenMultiSinkTablesBoundMultiSources() throws IOException {
```
In line with the difference of schema comment here - https://github.com/apache/hudi/pull/4925#issuecomment-1055454800, I feel it will be better to modify this test case to include the below scenario ->
1. sink_table1 bound to 2 topics
2. sink_table2 bound to a single topic
3. for sink_table1, let us have different schemas for the 2 source topics, I do not intend to have big differences in their schemas, absence of a single field in one of the schemas should suffice. Let us show the use case of matching the schemas to the target schema as discussed in the quoted comment. 

This should help users in understanding the kind of use cases covered in this PR. Later on we can either update the existing blog or write a new one to document this feature properly for everyone. Let me know in case of any doubts.

pratyakshsharma replied:
Once the suggested changes are incorporated, we can actually remove `testMultiTableExecutionWithKafkaSourceWhenOneSinkTableBoundMultiSources` and have a single test case covering the entire scenario.

pratyakshsharma reviewed (changes requested):
Thank you for the explanation @watermelon12138 . I feel the test case needs to modify as suggested which will cover the different scenarios possible. 

Once these changes are done, we should be good to land.

watermelon12138 commented:
@pratyakshsharma Hello, I've modified the test case to cover scene you suggested. The test case I added has passed. Please check whether there are other problems in the test case. Thank you very much!

xushiyan commented:
@watermelon12138 @pratyakshsharma are we able to land this in next few days for 0.11.0 ? or we push to 0.12

watermelon12138 commented:
@xushiyan
Thank you for your attention to this, and I hope to land this in next few days for 0.11.0, I'm debugging UT.

pratyakshsharma commented:
@watermelon12138 Still working on this? If you are facing any challenges or something is not clear, you can share with us and we can probably help.
